lib/input/sftp.go

```diff
@@ -0,0 +1,495 @@
+		}
+		if object, err = s.getObjectTarget(ctx); err != nil {
+			return
+		}
+	}
+
+	return sftpMsgFromPart(object, p), func(rctx context.Context, res types.Response) error {
+		return scnAckFn(rctx, res.Error())
+	}, nil
+}
+
+func sftpMsgFromPart(p *sftpPendingObject, part types.Part) types.Message {
+	msg := message.New(nil)
+	msg.Append(part)
+
+	meta := msg.Get(0).Metadata()
+
+	meta.Set("sftp_file_path", p.target.key)
+
+	return msg
+}
+
+func (s *SFTP) initSFTPConnection() error {
+	// create sftp client and establish connection
+	server := &SFTPServer{
+		Host: s.conf.Server,
+		Port: s.conf.Port,
+	}
+
+	certCheck := &ssh.CertChecker{
+		IsHostAuthority: hostAuthCallback(),
+		IsRevoked:       certCallback(server),
+		HostKeyFallback: hostCallback(server),
+	}
+
+	addr := fmt.Sprintf("%s:%d", s.conf.Server, s.conf.Port)
+	config := &ssh.ClientConfig{
+		User: s.conf.Credentials.Username,
+		Auth: []ssh.AuthMethod{
+			ssh.Password(s.conf.Credentials.Secret),
+		},
+		HostKeyCallback: certCheck.CheckHostKey,
+	}
+
+	var conn *ssh.Client
+	var err error
+	connectionAttempts := 0
+	for {
+		connectionAttempts++
+		conn, err = ssh.Dial("tcp", addr, config)
+		if err != nil {
+			connectionErrorsCounter := s.stats.GetCounter("connection_errors")
+			connectionErrorsCounter.Incr(1)
+			s.log.Errorf("Failed to dial: %s", err.Error())
+			if connectionAttempts >= s.conf.MaxConnectionAttempts {
+				s.log.Errorf("Failed to connect after %i attempts, stopping", connectionAttempts)
+				return errors.New("failed to connect to SFTP server")
+			}
+			var sleepDuration time.Duration
+			if sleepDuration, err = time.ParseDuration(s.conf.RetrySleepDuration); err != nil {
+				return fmt.Errorf("failed to parse retry sleep duration: %v", err)
+			}
+			time.Sleep(sleepDuration)
+		} else {
+			break
+		}
+	}
+
+	client, err := sftp.NewClient(conn)
+
+	if err != nil {
+		clientErrorsCounter := s.stats.GetCounter("client_errors")
+		clientErrorsCounter.Incr(1)
+		s.log.Errorf("Failed to create client: %s", err.Error())
+	}
+
+	s.client = client
+
+	return err
+}
+
+func (s *SFTP) getObjectTarget(ctx context.Context) (*sftpPendingObject, error) {
+	if s.object != nil {
+		return s.object, nil
+	}
+
+	target, err := s.keyReader.Pop(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = s.client.Stat(target.key)
+	if err != nil {
+		target.ackFn(ctx, err)
+		return nil, fmt.Errorf("target file does not exist: %v", err)
+	}
+
+	file, err := s.client.Open(target.key)
+	if err != nil {
+		target.ackFn(ctx, err)
+		return nil, err
+	}
+
+	object := &sftpPendingObject{
+		target: target,
+		obj:    file,
+	}
+
+	if object.scanner, err = s.objectScannerCtor(target.key, file, target.ackFn); err != nil {
+		target.ackFn(ctx, err)
+		return nil, err
+	}
+
+	s.object = object
+	return object, nil
+}
+
+// SFTPServer contains connection data for connecting to an SFTP server
+type SFTPServer struct {
```

Comment:
Since this is duplicated for the output as well it might be worth pulling it out into a shared package similar to https://github.com/Jeffail/benthos/tree/master/internal/service/redis